### PR TITLE
Test for GitHub pull requests

### PR DESCRIPTION
**Pull request by kwhetstone**
First test case for scanning a GitHub repo. This mocks the GitHub responses so no actual connection to GitHub is needed; testing can be done in isolation. Future work will update these mocks so they won't be out of date, and can catch errors before going out as a release. Test coverage will expand to other classes   and utilities soon.  The tests here are pretty simplistic but it can be built up further.

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
Supersedes #55 I suppose.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        LogTaskListener logListener = new LogTaskListener(Logger.getLogger(getClass().getName()), Level.FINE);
+
+        GitHubSCMSource ghss = new GitHubSCMSource("testSource", "http://localhost:8089/", null, null, "testOwner", "testRepo");
+        Map<SCMHead, SCMRevision> results = getResults(ghss);
+        SCMHead master = getMaster(results.keySet());
+        SCMRevision mRev = results.get(master);
+
+        SCM scm = ghss.build(master, mRev);
+        assertNotNull("SCM exists", scm);
+        assertEquals("hudson.plugins.git.GitSCM", scm.getType());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java]: This is not testing what `build` actually returns, other than it does not throw an exception. So we probably need some way of swapping in a local (`file`-protocol) URL from `GitSampleRepoRule` to do full tests.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public WireMockRule wireMockRule = new WireMockRule(8089);
+
+
+    @Rule
+    public JenkinsRule jrule = new JenkinsRule();
+
+
+    @Test
+    public void testRetrieve() throws Exception {
+        GitHubSCMSource ghss = new GitHubSCMSource("testSource", "http://localhost:8089/", null, null, "testOwner", "testRepo");
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java]: So no credentials.

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java]: Perhaps we need to create an actual access token, test everything with real requests and responses, then invalidate the token before pushing.

**Comment by jglick** [comment]
So the three things I think we need to add:

* Some system for testing with credentials, once we figure out the policy for temporary access tokens. A special testing organization with private repositories?
* A way to rerun one test, or all tests, against the live server, so we can see if behavioral changes on GitHub’s side broke something; and get a diff of requests and responses.
* A way to list mock requests & responses which were *not* used during `mvn test`, so we can “garbage-collect” them from the source tree.

**Comment by jglick** [comment]
What about tests of caching systems using `If-Modified-Since` and (IIRC) 304 responses?

**Comment by kwhetstone** [comment]
Built the test-harness and installed locally (I don't have the ability to deploy a snapshot into staging), and the tests passed.  This PR is now dependent on jenkinsci/jenkins-test-harness#39.
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running InjectedTest
Running org.jenkinsci.plugins.github_branch_source.GitHubSCMSourceTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 76.815 sec - in org.jenkinsci.plugins.github_branch_source.GitHubSCM
SourceTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 78.512 sec - in InjectedTest

Results :

Tests run: 14, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java [hunk 3]
```diff
@@ -17,16 +17,17 @@
...
 import static org.junit.Assert.*;
-import static org.hamcrest.Matchers.contains;
 
 public class GitHubSCMSourceTest {
-    @Rule
-    public WireMockRule wireMockRule = new WireMockRule(8089);
+    WireMockRuleFactory wmrf = new WireMockRuleFactory();
 
+    @Rule
+    public WireMockRule wireMockRule = wmrf.getRule(8089);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java]: I guess you could inline `wmrf`, right?

```java
@Rule
public WireMockRule wireMockRule = new WireMockRuleFactory().getRule(8089);
```

Not clear if there is some use case for keeping it separate—to listen on multiple ports at once?

File: pom.xml [hunk 4]
```diff
@@ -81,6 +81,17 @@
...
+        <dependency>
+            <groupId>org.jenkins-ci.main</groupId>
+            <artifactId>jenkins-test-harness</artifactId>
+            <version>2.18-SNAPSHOT</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>19.0</version>
```
Comment (jglick) [inline pom.xml]: Why is this here? It will not work—Jenkins core bundles a version of Guava, which you cannot easily override (you would have to set a plugin-first class loader, which is rather dangerous). When you add a non-plugin `compile`-scoped dependency like this you are _bundling_ that library into your plugin.

**Comment by stephenc** [comment]
recommend closing in favour of #93

**Comment by kwhetstone** [comment]
I agree. Since the wiremock change hasn't been delivered in the jenkins-test-harness, this would be blocked by that anyway. I'm for closing.
